### PR TITLE
linux_latest: 5.17.9 -> 5.18

### LINKED ISSUE #174480: Linux 5.18 fails to build
### Steps To Reproduce
1. Run `nix build github:johnrichardrinehart/johnos#nixosConfigurations.simple-live-iso.pkgs.JohnOS-kernel.kernel`

### Build log
Tail of the build log
```
QUESTION: Memtest, NAME: MEMTEST, ALTS: Y/n/?, ANSWER: y
GOT: y
QUESTION: Microsoft Hyper-V driver testing, NAME: HYPERV_TESTING, ALTS: N/y/?, ANSWER: 
GOT: 
GOT: #
GOT: # configuration written to .config
GOT: #
GOT: make[1]: Leaving directory '/build/linux-5.18/build'
GOT: make: Leaving directory '/build/linux-5.18'
warning: unused option: ASHMEM
error: unused option: DEBUG_INFO
warning: unused option: DEBUG_INFO_BTF
warning: unused option: DEBUG_STACKOVERFLOW
warning: unused option: EXT4_ENCRYPTION
warning: unused option: F2FS_FS_ENCRYPTION
error: unused option: NFSD_V3
warning: unused option: RCU_PERF_TEST
warning: unused option: XEN_SCRUB_PAGES
warning: unused option: XEN_SELFBALLOONING
warning: unused option: XEN_STUB
warning: unused option: XEN_TMEM
```

The build failure seems related to these unused options. I tried removing the corresponding lines (in a local directory - so I changed my `input.nixpkgs` url to the filesystem path). in `/pkgs/os-specific/linux/kernel/common-config.nix` but it didn't fix anything.

### Additional context

It seems that kernel 5.18 phased out some options. Instead of warning us, there's an explicit error. I have no idea how to fix this since `extraConfig` is great for adding lines to the kernel config, but not great at removing them.

### Notify maintainers

@xaverdh @copumpkin @teto (I'm not exactly sure who to notify about this so I just kind of scanned `git blame`s of `common-config.nix` and `top-level/linux-kernels.nix`. Please let me know if there's someone more appropriate to notify.)

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

Not sure how to do this for this specific package (no `default.nix`).

### PR DESCRIPTION
**Pull request by alyssais**
###### Description of changes

NSFD_V3 is now always enabled, and enabling debug info now requires selecting a DWARF version instead of just setting DEBUG_INFO=y.

I'm doing some further work to update some out-of-tree modules, mark others as broken, etc. (see my last few PRs), but will not be blocking this PR for anything out-of-tree.

Fixes https://github.com/NixOS/nixpkgs/issues/174480.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
@OfBorg test login

**Comment by SuperSandro2000** [comment]
Do you know by chance if https://cateee.net/lkddb/web-lkddb/DEBUG_INFO.html has already 5.18 in it?

Does ZFS already work with 5.18 or did I miss your PR about it?

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 1]
```diff
@@ -38,8 +38,9 @@ let
       # Necessary for BTF
       DEBUG_INFO                = mkMerge [
         (whenOlder "5.2" (if (features.debug or false) then yes else no))
-        (whenAtLeast "5.2" yes)
+        (whenBetween "5.2" "5.18" yes)
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/kernel/common-config.nix:41]: I dug up https://salsa.debian.org/kernel-team/linux/-/commit/6834606ba61e797a3bfd88731a0e8924b1566170 which means we only need to set DEBUG_INFO_DWARF_TOOLCHAIN_DEFAULT to y on 5.18.

Comment (alyssais) [reply pkgs/os-specific/linux/kernel/common-config.nix:41]: Isn't that what I'm doing here?  `whenBetween "5.2" "5.18"` means including 5.2, but not including 5.18.  So on 5.18 we only set DEBUG_INFO_DWARF_TOOLCHAIN_DEFAULT, not DEBUG_INFO.

**Comment by alyssais** [comment]
> Does ZFS already work with 5.18 or did I miss your PR about it?

I don't know.  But it doesn't matter because it's marked broken for 5.18, and `zfs.latestCompatibleLinuxPackages` is set to 5.17.

File: pkgs/os-specific/linux/kernel/common-config.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/os-specific/linux/kernel/common-config.nix:41]: Oh, yeah, somehow missed that.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by johnrichardrinehart** [comment]
@alyssais Thanks for doing this work! I couldn't figure out how to do it myself!

**Pull request merged**